[WIP] FIND_NODE iteration

treygriffith opened the pull request:
@bookchin here's a possible implementation of `iterativeFindNode` that matches how I read the [spec](http://xlattice.sourceforge.net/components/protocol/kademlia/specs.html).

I took some of the structure and conventions from #109.

It still needs:

- [x] Additional unit and e2e tests for full spec compliance
- [x] Updated `iterativeFindValue` implementation to match `iterativeFindNode`

If you like this approach, I can continue with the above, or if you prefer, you can fold in some of these concepts into your WIP on #109.

treygriffith commented on lib/node-kademlia.js:
```diff
@@ -162,28 +162,101 @@ class KademliaNode extends AbstractNode {
+        // NB: If we have reached at least K active nodes, or haven't found a
+        // NB closer node, even on our finishing trip, return to the caller
+        // NB: the K closest active nodes.
+        if (maxActivePeers || (!foundCloserNode() && !continueLookup)) {
```
The spec is somewhat ambiguous on this point, but in my view if we found a closer node when `continueLookup === false`, we should continue the search if we don't have `maxActivePeers` yet.

bookchin replied:
Yeah, the spec seems to leave this up to implementors. I agree that if we haven't reached our max contacts and there are potentially closer nodes, we should keep going.

bookchin commented:
Awesome! If you're on a roll, feel free to continue on and I'll close my other PR. I'm recovering from a wrist injury, so I can only hack on this stuff for short periods of time right now and I'd rather not let that slow down this work especially since it's clear you have a strong grasp on what needs to be done. 

Just poke me when you're ready for me to review (and don't forget to add yourself to the contributors section of the `package.json`)!

treygriffith commented on lib/node-kademlia.js:
```diff
@@ -68,7 +69,7 @@ class KademliaNode extends AbstractNode {
     async.series([
       (next) => this.iterativeFindNode(this.identity.toString('hex'), next),
-      (next) => this.refresh(this.router.getBucketsBeyondClosest()[0], next)
+      (next) => this.refresh(this.router.getClosestBucket() + 1, next)
```
Based on my reading of the spec, refreshes should address all buckets beyond the closest neighbor. Is there a reason this used `getBucketsBeyondClosest` before?

treygriffith commented on lib/node-kademlia.js:
```diff
@@ -371,11 +293,10 @@ class KademliaNode extends AbstractNode {
       const needsRefresh = lastBucketLookup + constants.T_REFRESH <= now;
 
-      if (bucketHasContacts && needsRefresh) {
+      if (needsRefresh) {
```
I changed this to refresh any bucket that's old, rather than only refreshing buckets that already have contacts. Is there a reason that it was set up to only refresh buckets with contacts already in them?

treygriffith commented:
@bookchin I think it's ready for review.

I added a new data structure, `ContactList` to try to make the rules about collecting nodes a bit clearer.

I refactored `iterativeFindValue` and `iterativeFindNode` into a single function, since `iterativeFindNode` is a kind of special case of `iterativeFindValue` that doesn't return early.

I changed the implementation of `refresh` to refresh all buckets instead of just ones that already have nodes - I don't see from the spec why it would be implemented that way.

The testing is a bit sluggish now, particularly on `join`. Any ideas for how to speed that up?

bookchin commented:
Yeah, the e2e tests are particularly slow  - which correlates directly to the number of nodes in the test. I think 32 is hardcoded. Ultimately it'd be really nice to fork all those into child processes, but that's just not how I originally implemented it. 

Thanks for all your work so far! I'll review this over the next couple days. :)
